File: app/api/routes/projects.py
from __future__ import annotations

import uuid

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.schemas.project import ProjectCreate, ProjectListResponse, ProjectResponse
from app.models.database import get_db
from app.services import project_service

router = APIRouter(prefix="/api/projects", tags=["projects"])

# Placeholder user ID until auth is implemented
PLACEHOLDER_USER_ID = uuid.UUID("00000000-0000-0000-0000-000000000001")


@router.post("", response_model=ProjectResponse, status_code=201)
async def create_project(
    body: ProjectCreate,
    db: AsyncSession = Depends(get_db),
) -> ProjectResponse:
    try:
        project = await project_service.create_project(
            db, PLACEHOLDER_USER_ID, body.name, body.description, body.template_key
        )
    except KeyError as e:
        raise HTTPException(status_code=400, detail=str(e))
    return ProjectResponse.model_validate(project)


@router.get("", response_model=ProjectListResponse)
async def list_projects(
    db: AsyncSession = Depends(get_db),
) -> ProjectListResponse:
    projects = await project_service.list_projects(db, PLACEHOLDER_USER_ID)
    return ProjectListResponse(projects=[ProjectResponse.model_validate(p) for p in projects])


@router.get("/{project_id}", response_model=ProjectResponse)
async def get_project(
    project_id: uuid.UUID,
    db: AsyncSession = Depends(get_db),
) -> ProjectResponse:
    project = await project_service.get_project(db, project_id)
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")
    return ProjectResponse.model_validate(project)


@router.delete("/{project_id}", status_code=204)
async def delete_project(
    project_id: uuid.UUID,
    db: AsyncSession = Depends(get_db),
) -> None:
    deleted = await project_service.delete_project(db, project_id)
    if not deleted:
        raise HTTPException(status_code=404, detail="Project not found")

File: frontend/src/lib/api/projects.ts
import { apiFetch } from "../api-client";

export interface Project {
  id: string;
  name: string;
  description: string | null;
  template_key: string;
  status: string;
  chat_phase: string | null;
  selected_doc_types: string[] | null;
  created_at: string;
}

export interface ProjectListResponse {
  projects: Project[];
}

export function fetchProjects(): Promise<ProjectListResponse> {
  return apiFetch("/api/projects");
}

export function fetchProject(id: string): Promise<Project> {
  return apiFetch(`/api/projects/${id}`);
}

export function createProject(data: {
  name: string;
  description?: string;
  template_key?: string;
}): Promise<Project> {
  return apiFetch("/api/projects", {
    method: "POST",
    body: JSON.stringify(data),
  });
}

export function createProjectFromChat(data: {
  initial_prompt: string;
  name?: string;
}): Promise<{ id: string; project_id: string; content: string }> {
  return apiFetch("/api/projects/from-chat", {
    method: "POST",
    body: JSON.stringify(data),
  });
}

export function deleteProject(id: string): Promise<void> {
  return apiFetch(`/api/projects/${id}`, { method: "DELETE" });
}
